### PR TITLE
add error-handling exercise

### PR DESCRIPTION
**Pull request by A-O-Emmanuel**
Implemented the Error Handling practice exercise with solution, tests, and documentation

### CODE REVIEW

File: exercises/practice/error-handling/.meta/proof.ci.js
Comment (SleeplessByte) [inline exercises/practice/error-handling/.meta/proof.ci.js:1]: The contents are missing here, that's why the CI is failing!

File: exercises/practice/error-handling/error-handling.js
Comment (SleeplessByte) [inline exercises/practice/error-handling/error-handling.js:1]: All the other exercises have a "stub" (initial code) that aids with implementation. Can you check the other concept exercises how we do that and add it here? It probably needs to define `processString` and merely raise an error to implement the function.

File: exercises/practice/error-handling/error-handling.spec.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    } catch {
+      /*
+                intentionally left empty,
+                I expext this call to throw,
+                but only care about verifying that the finally block is executed
+                and clean up message logged.
+            */
+    }
+
+    expect(console.log).toHaveBeenCalledWith('Resource cleaned up');
```
Comment (SleeplessByte) [inline exercises/practice/error-handling/error-handling.spec.js]: I don't think this will work with the [Browser Test Runner](https://github.com/exercism/js-test-runner-clientside), but the idea is not wrong. I can help come up with an easy alternative to fix this.

**Review by SleeplessByte — CHANGES_REQUESTED** [review]

File: exercises/practice/error-handling/.meta/config.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
+  "authors": ["Obinna Obi-Akwari"],
```
Comment (BNAndras) [inline exercises/practice/error-handling/.meta/config.json]: This should be your GitHub user name, not what I’m guessing is your actual name.

File: exercises/practice/error-handling/error-handling.spec.js (hunk 1 above)
Comment (A-O-Emmanuel) [reply exercises/practice/error-handling/error-handling.spec.js]: okay, thank you

File: exercises/practice/error-handling/error-handling.js
Comment (A-O-Emmanuel) [reply exercises/practice/error-handling/error-handling.js:1]: Alight, I'll do that

File: exercises/practice/error-handling/.meta/config.json (hunk 2 above)
Comment (A-O-Emmanuel) [reply exercises/practice/error-handling/.meta/config.json]: okay, i'll add it, thank you

File: exercises/practice/error-handling/error-handling.spec.js
Comment (BNAndras) [inline exercises/practice/error-handling/error-handling.spec.js:1]: Generally, tests after the first one should be skipped by default. The online test runner will run all the tests, but that allows students working offline to practice TDD and solve tests one at a time.

You’re using `it` here, but the convention on this track is to use `test` and the corresponding `xtest`. See https://github.com/exercism/javascript/blob/main/exercises/practice/leap/leap.spec.js as an example.

File: exercises/practice/error-handling/error-handling.spec.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    expect(() => processString(42)).toThrow(TypeError);
+    expect(() => processString(42)).toThrow('input must be a string');
+  });
+
+  it('throws Error with EmptyStringError if string is empty', () => {
+    expect(() => processString('')).toThrow(Error);
+    expect(() => processString('')).toThrow('EmptyStringError');
+  });
+
+  it('always logs cleanup message', () => {
```
Comment (BNAndras) [inline exercises/practice/error-handling/error-handling.spec.js]: This test checks if a cleanup message is logged after an empty string is provided. That isn’t the same as checking if a cleanup message is always logged.

File: exercises/practice/error-handling/error-handling.spec.js
Comment (BNAndras) [reply exercises/practice/error-handling/error-handling.spec.js:1]: Also would it make sense to use custom exceptions here? Then you don’t need to check both the type and message.

File: exercises/practice/error-handling/error-handling.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export const processString = (input) => {
+    //TODO: implement this
+    //should throw Error if input is not a string
```
Comment (BNAndras) [inline exercises/practice/error-handling/error-handling.js]: This line isn’t correct. The test suite is looking for a specific error type.

File: exercises/practice/error-handling/error-handling.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+    //should throw Error if input is an empty string
```
Comment (BNAndras) [inline exercises/practice/error-handling/error-handling.js]: This implies any Error can be thrown with any message, but that’s not true.

File: exercises/practice/error-handling/error-handling.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+    //should return the uppercase version of the string if valid
```
Comment (BNAndras) [inline exercises/practice/error-handling/error-handling.js]: What’s considered valid here? This line implies by nearness that not all strings are valid, but it’s not clear what is valid.

File: exercises/practice/error-handling/error-handling.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+    //should use finally to log cleanup message
```
Comment (BNAndras) [inline exercises/practice/error-handling/error-handling.js]: Cleanup of what exactly?

**Review by BNAndras — COMMENTED** [review]
I’m not a track maintainer here, just a friendly maintainer from elsewhere on Exercism. I made some more general comments, but whatever SleeplessBytes and Cool-Kat say is the final word.  :)

File: exercises/practice/error-handling/error-handling.spec.js
Comment (A-O-Emmanuel) [reply exercises/practice/error-handling/error-handling.spec.js:1]: Any advice and recommendation is good for me, thank you I will look at everything you said

File: exercises/practice/error-handling/error-handling.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+export const processString = (input) => {
+  //TODO: implement this
+  //should throw TypeError if input is not a string
+  //should throw a general Error if input is an empty string
+  //should return the uppercase version of the string 'hello'
```
Comment (SleeplessByte) [inline exercises/practice/error-handling/error-handling.js]: This is part of the instructions so should not be part of our stub
```suggestion
```

File: exercises/practice/error-handling/error-handling.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+export const processString = (input) => {
+  //TODO: implement this
```
Comment (SleeplessByte) [inline exercises/practice/error-handling/error-handling.js]: In all other stubs we `throw new Error('...')` (I don't recall the exact error message). We probably want to use that!

File: exercises/practice/error-handling/.meta/proof.ci.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+export const processString = (input) => {
+  if (typeof input !== 'string') {
+    throw new TypeError();
+  }
+  if (input === '') {
+    throw new Error();
```
Comment (SleeplessByte) [inline exercises/practice/error-handling/.meta/proof.ci.js]: We almost always want a message in an error. We don't need to test the exact message, but throwing an empty error is non-idiomatic.

File: exercises/practice/error-handling/.meta/proof.ci.js [hunk 11]
```diff
@@ -0,0 +1,9 @@
+export const processString = (input) => {
+  if (typeof input !== 'string') {
+    throw new TypeError();
```
Comment (SleeplessByte) [inline exercises/practice/error-handling/.meta/proof.ci.js]: We almost always want a message in an error. We don't need to test the exact message, but throwing an empty error is non-idiomatic.

**Review by SleeplessByte — COMMENTED** [review]
What's there is almost done. The only issue I have is that this isn't actually *handling* any error!

File: exercises/practice/error-handling/error-handling.js (hunk 9 above)
Comment (A-O-Emmanuel) [reply exercises/practice/error-handling/error-handling.js]: Okay

File: exercises/practice/error-handling/.meta/proof.ci.js (hunk 10 above)
Comment (A-O-Emmanuel) [reply exercises/practice/error-handling/.meta/proof.ci.js]: Alright, I'll add that

**Comment by A-O-Emmanuel** [comment]
Hi @SleeplessByte thanks for your review, sorry it took this long for me to reply,  just to clarify, when you say "The only issue I have is that this isn't actually handling any error!" do you mean that the function should include descriptive error messages rather than handling more complex error scenarios?

**Comment by SleeplessByte** [comment]
@A-O-Emmanuel no worries! We are not in a hurry.

Right now, the exercise wants you to throw errors. It doesn't want you to *handle* the error.

Handling errors could be:

```javascript
try { } catch { }
```

or

```javascript
promise.catch(...)
```

or

```javascript
try { } finally { }
```

or 

```javascript
promise.finally(...)
```

**Comment by A-O-Emmanuel** [comment]
Oh, that's true, I'll just have to revert back to the way I did it before. Thanks for the clarification, I've learnt something today, "throwing errors and handling errors" are two different things, yeah, it makes sense, thank you.

**Comment by SleeplessByte** [comment]
@Cool-Katt how do you feel about this?

**Comment by Cool-Katt** [comment]
I was kinda out of the loop on this one so I can review in more depth tomorrow, but preliminary assessment is that I would have liked to see a couple of more tests. 

I'll probably have to get up to speed first tho, to comment more

**Comment by A-O-Emmanuel** [comment]
Hi @Cool-Katt, okay I will add more tests, but I'll wait for your final assessment before adding more tests.

File: exercises/practice/error-handling/error-handling.js [hunk 12]
```diff
@@ -0,0 +1,8 @@
+//
+// This is only a SKELETON file for the 'Error handling' exercise. It's been provided as a
+// convenience to get you started writing code faster.
+//
+
+export const processString = (input) => {
+  throw new Error('Remove this line and implement the function');
```
Comment (Cool-Katt) [inline exercises/practice/error-handling/error-handling.js:7]: the way this currently is, it'll pass one of the tests by default, without the student changing anything.

File: exercises/practice/error-handling/error-handling.spec.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
+import { describe, expect, test, xtest } from '@jest/globals';
+import { processString } from './error-handling';
+
+describe('Error Handling', () => {
+  test('throws TypeError if input is not a string', () => {
+    expect(() => processString(42)).toThrow(TypeError);
+  });
+
+  xtest('throws Error message if string is empty', () => {
```
Comment (Cool-Katt) [inline exercises/practice/error-handling/error-handling.spec.js]: Idk if i'm vibing with this test specifically, maybe we need to add some handling for the empty string case, instead of throwing error.

In any case, more tests are needed, like for example do we want the tests to pass with any generic `Error` type or do we want to allow only certain types?

File: exercises/practice/error-handling/.docs/instructions.md [hunk 14]
```diff
@@ -0,0 +1,8 @@
+# Instructions
+
+Implement various kinds of error handling and resource management.
+
+An important point of programming is how to handle errors and close resources even if errors occur.
+
+This exercise requires you to handle various errors.
+Because error handling is rather programming language specific you'll have to refer to the tests for your track to see what's exactly required.
```
Comment (Cool-Katt) [inline exercises/practice/error-handling/.docs/instructions.md]: this line probably needs to be removed, or at least changed because it's generic from the `problem-spec` repo

File: exercises/practice/error-handling/.docs/introduction.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+- Handle and catch errors using a try..catch block
+
+
+Implement the processString function using a `try…catch` block.
+
+Inside the `try` block:
+
+- If the input is not a string, throw a `TypeError`.
+
+- If the input is an empty string, throw a generic `Error`.
```
Comment (Cool-Katt) [inline exercises/practice/error-handling/.docs/introduction.md]: how about on empty string, instead of throwing, it returns `'INVALID'` or `null/undefined` so we can add some handling too.

we can also explicitly forbid the generic `Error` type to add more complexity

**Comment by Cool-Katt** [comment]
Right, sorry folks, real life got in the way. Anyway.

The way I see it, currently its a good effort but i'd maybe like to see some more restrictions/explicit handling scenarios. It also definitely needs more tests and maybe a rewrite of the instructions.

**Comment by A-O-Emmanuel** [comment]
Okay @Cool-Katt, thanks, just got the chance to look at your comments, been really occupied, but I should be free ending of next week to start implementing your suggestions, thank you.

**Comment by A-O-Emmanuel** [comment]
hi @Cool-Katt I've made changes and added more tests, please review, and sorry for the delay.

**Comment by A-O-Emmanuel** [comment]
Hi @Cool-Katt have you been able to review the changes I made?

**Comment by Cool-Katt** [comment]
@A-O-Emmanuel Terribly sorry for the long delay! It shouldn't have taken me 5 months to rely, yikes 😬

I still stand by my previous suggestion, which is:
I don't like to just throw a generic `Error` because that way the tests will pass by default without having changed the code at all, so I propose instead to throw a `RangeError` for the two cases that test length and maybe a `SyntaxError` for the case with mixed letters/numbers. We don't have to test for a specific message, but we can test for a specific type of error and weather or not a message was included at all (which it should be).
Also, there should be a test that fails in case a generic `Error` is thrown.

Let me know if you still want to work on this?

**Comment by A-O-Emmanuel** [comment]
@Cool-Katt thanks for the comments, I will still work on it. by this weekend, I will create the commit with your suggestion, also, it's okay, I just thought you were busy. Thank you.

**Comment by A-O-Emmanuel** [comment]
@Cool-Katt I've just push the changes you requested, please review

**Comment by A-O-Emmanuel** [comment]
Hi @Cool-Katt @BNAndras @SleeplessByte @joshgoebel can anyone provide any feedback for this?

File: exercises/practice/error-handling/.docs/introduction.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+Inside the `try` block:
+
+- If the input is not a string, throw a `TypeError`.
+
+- If the input is an empty string, return `null`.
+
+- If input length is greater than 100, throw a `RangeError`.
+
+- If input length is less than 10, throw a `RangeError`.
```
Comment (Cool-Katt) [inline exercises/practice/error-handling/.docs/introduction.md]:
```suggestion
- If input length is greater than 100, or less than 10, throw a `RangeError`
```

File: exercises/practice/error-handling/.docs/introduction.md [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+- If input contains a mix of letters and numbers, throw a `SyntaxError`.
+
+- Otherwise, return the input in `uppercase`.
+
```
Comment (Cool-Katt) [inline exercises/practice/error-handling/.docs/introduction.md:21]:
```suggestion

Don't forget to attach appropriate error messages then you throw! An informative and well structured error message can save you hours of debugging.
```

File: exercises/practice/error-handling/error-handling.spec.js [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+      let error;
+
+      try {
+        processString(input);
+      } catch (err) {
+        error = err;
+      }
+
+      expect(error).toBeInstanceOf(Error);
+      expect(error.constructor).not.toBe(Error);
```
Comment (Cool-Katt) [inline exercises/practice/error-handling/error-handling.spec.js]: We only need this check of the three, but testing for a message is good in any case because we don't ant people to circumvent sending one.

Also, i feel like this should be the first test, followed by all the tests for the various error types and finally  the two passing cases that test the functionality

File: config.json [hunk 19]
```diff
@@ -2770,6 +2770,14 @@
...
         "prerequisites": [],
         "difficulty": 2
+      },
+      {
+        "slug": "error-handling",
+        "name": "Error Handling",
+        "uuid": "de1c75f2-2461-4347-b5ca-b3cfaafe4d79",
+        "practices": [],
+        "prerequisites": [],
+        "difficulty": 1
```
Comment (Cool-Katt) [inline config.json]: set an appropriate difficulty level. You can cross reference on the site to see what level exercises show up al easy, medium or hard. (I don't remember the scale, sry)

File: exercises/practice/error-handling/error-handling.spec.js [hunk 20]
```diff
@@ -0,0 +1,72 @@
+import { describe, expect, test, xtest } from '@jest/globals';
+import { processString } from './error-handling';
+
+describe('Error Handling', () => {
+  xtest('throws TypeError if input is not a string', () => {
```
Comment (Cool-Katt) [inline exercises/practice/error-handling/error-handling.spec.js:28]: the first test in the stack should be unskipped, all the other ones should be skipped. 
```suggestion
  test('throws TypeError if input is not a string', () => {
```

**Review by Cool-Katt — CHANGES_REQUESTED** [review]
The PR for adding a new exercise should ideally only touch the files relevant to the new exercise being added, so please don't change any of the other files (even if it's just formatting, we have an auto formatter for this).

Also, see comments bellow

**Comment by Cool-Katt** [comment]
@A-O-Emmanuel Overall good stuff. Just needs a few very minor tweaks and we can **finally** merge this

File: config.json (hunk 19 above)
Comment (A-O-Emmanuel) [reply config.json]: Okay, I'll do this later today, thank you

File: exercises/practice/error-handling/error-handling.spec.js (hunk 18 above)
Comment (A-O-Emmanuel) [reply exercises/practice/error-handling/error-handling.spec.js]: Okay, thank you, and for the other files, sorry about that, It was mistake, I didn.t even know the formater ran for everyfile.

**Comment by A-O-Emmanuel** [comment]
HI @Cool-Katt I've implemented the changes you request, please check if they are up to the standard you want, thank you

**Comment by A-O-Emmanuel** [comment]
okay, doing that now

**Comment by Cool-Katt** [comment]
@A-O-Emmanuel that was for the bot, but i got a typo. You don't have to do anything

**Review by Cool-Katt — APPROVED** [review]
LGTM!

**Comment by A-O-Emmanuel** [comment]
Okay, thank you

**Comment by A-O-Emmanuel** [comment]
@Cool-Katt so it will be merged? or any other thing i need to do?

**Comment by Cool-Katt** [comment]
@A-O-Emmanuel Nothing else for you to do right now. 

If @SleeplessByte approves as well (since they requested some changes) we can merge this.

**Comment by A-O-Emmanuel** [comment]
Okay, but I think @SleeplessByte was waiting for you to approve, let me send him a message

**Review by SleeplessByte — APPROVED** [review]
